GROUP BY implementation

dt opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3494)

<!-- Reviewable:end -->

dt committed:
pull encode DTuple out into own helper

dt committed:
aggregate impl map: switch instances to funcs

Also un-export impl methods.

dt committed:
normalize group-by expressions to match render expressions

dt committed:
GROUP BY implementation

Add buckets of values to aggregateFuncs, and capture grouped qvalues.
Render buckets into rows of a valuesNode on first call to Next.

Note: ORDER BY (with GROUP BY) still unsupported (as is HAVING).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=3494) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

petermattis commented:
Nice.

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/group.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qT9g1FTu9LMzYcC92:-K5qT9g1FTu9LMzYcC93:-2019735981) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L38)):*
`s/addRenderer/addRender/g`

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5qTHiP3g5z8y8Tpuma:829090311) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
I think it would be safer to check for `err != nil` first. Otherwise, you might return an error but set `aggregation` to `false` and thus accidentally ignore the error.

---

*[sql/group.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTY_-X3pb-mpKTBuz:-K5qTY_-X3pb-mpKTBv-:-1882369721) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L70)):*
`s/just return/returns/g`

---

*[sql/group.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTm5-auZG8PfX7lx3:-K5qTm5-auZG8PfX7lx4:-1041087360) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L84)):*
I thought the aggregation functions correctly handled no rows being fed to them.  Maybe this works differently with grouping.

---

*[sql/group.go, line 127 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTzuETTbLQe1KFGaA:-K5qTzuETTbLQe1KFGaB:255493526) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L127)):*
`s/currnetBucket/currentBucket/g`

---

*[sql/group.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qWRYTZ2ZyTdTzkfCT:-K5qWRYTZ2ZyTdTzkfCU:1188701615) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L177)):*
Add a TODO (or file an issue) that we can avoid bucketing if `groupedValues` are in the correct order. This will require changing `desiredAggregateOrdering` similar to how the desired ordering is computed for ORDER BY.

---

*[sql/group.go, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qUDw-k1zf6nhXuHKP:-K5qUDw-k1zf6nhXuHKQ:1256333074) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L178)):*
`s/values/values./g`

---

*[sql/testdata/aggregate, line 301 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qVlydOdV0Gsr9z25P:-K5qVlydOdV0Gsr9z25Q:1841300430) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/testdata/aggregate#L301)):*
Does `FUNC(DISTINCT ...)` interact with `GROUP BY`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-:-K5qWtqYDk7flOD5JnCe:-490100116)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5qY2KMLo-sE719VZSH:1733891597) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
So right now, err _should_ be ignored if aggregation is false: it is non-nil if any qvalues aren't aggregated, but that is only a real error condition if aggregation is in use. It might make sense to filter it inside `checkAggregateExprs` ?

---

*[sql/group.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTm5-auZG8PfX7lx3:-K5qZ4OLsS9BBLa58Qd6:1968696203) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L84)):*
it isn't the functions themselves but rather the rendering of rows that needs this:
when _not_ grouping, postgres returns a single row of nulls when nothing is aggregated -- when grouping, it returns 0 rows. This flag is to force the single empty row to be inserted in that first case.

---

*[sql/group.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qWRYTZ2ZyTdTzkfCT:-K5qZaSPamru2W0G2qkc:-1196295880) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L177)):*
Yeah, ORDER BY doesn't work at all right now with group -- I was going to look at that next. I'll file an issue if there isn't one.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/group.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qWRYTZ2ZyTdTzkfCT:-K5qaxMzffzarvJXohBi:-1129220099) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L177)):*
I'm referring to something else here: we can avoid using the bucket map if we are able to scan the rows and have them ordered by the group by key. For example, if you are grouping on the primary key and are scanning the primary index the rows will already be grouped together.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/group.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qWRYTZ2ZyTdTzkfCT:-K5qboNMXQDqRC-LOk6i:1540530549) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L177)):*
Ah, I'll add a TODO in computeAggregates mentioning that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5qdV1F1gIU5-0xNBOe:1726303114) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
Yeah. As it is written, this looks error prone (even if it isn't).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[sql/group.go, line 38 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qj-aIa46pzsqnnTxm:-K5qj-aIa46pzsqnnTxn:-47299654) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L38)):*
s/ them/ to them/

---

*[sql/group.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjAJXlY36Ij4mZH9A:-K5qjAJXlY36Ij4mZH9B:1806185395) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L55)):*
I'd put this description (explaining how this works) in the method comment on `checkAggregateExprs` instead of here

---

*[sql/group.go, line 67 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjXg4ZY9l59msrERs:-K5qjXg4ZY9l59msrERt:-559757923) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L67)):*
put this inline above?

---

*[sql/group.go, line 73 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjLcrWAVy3CFe7lVI:-K5qjLcrWAVy3CFe7lVJ:-602585334) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L73)):*
s/function/functions/

---

*[sql/group.go, line 96 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjeJbrp5n_qgvxjvZ:-K5qjeJbrp5n_qgvxjv_:-490100116) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L96)):*
Nice.

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K5qkIxx_pVhHG7UvIJc:-111304024) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
delegate this to `values`?

---

*[sql/group.go, line 127 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjir6R5CUrpTXLAAx:-K5qjir6R5CUrpTXLAAy:649866302) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L127)):*
is this a pointer so that you can tell the difference between initialized and uninitialized? add a comment.

---

*[sql/group.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkOYnMoMtxcjsprY5:-K5qkOYnMoMtxcjsprY6:407358626) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L164)):*
instead of this (allocation), use `var scratch [64]byte`, then pass `scratch[:0]` below.

---

*[sql/group.go, line 278 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkyc0Ps8I-C4TUNvc:-K5qkyc0Ps8I-C4TUNvd:-742291377) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L278)):*
nit: this doesn't mutate the receiver - call it something like `new()` rather than `init`

---

*[sql/group.go, line 327 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qlJAN8wIA0ObZlRcY:-K5qlJAN8wIA0ObZlRcZ:-490100116) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L327)):*
Nice.

---

*[sql/group.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qlgfSd51uzbtPr2Zt:-K5qlgfSd51uzbtPr2Zu:-420893807) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L399)):*
Might be worth a TODO here, but this comment is great.

---

*[sql/group.go, line 471 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qlvMTsxgu223XIVWR:-K5qlvMULtjDA6f0MEbs:1748001933) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L471)):*
can this ever happen? when?

---

*[sql/group.go, line 528 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qm6hPk5aZNsGrz5wa:-K5qm6hPk5aZNsGrz5wb:826851976) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L528)):*
why is an uninitialized identAggregate equivalent to one initialized with NULL? add a comment.

---

*[sql/testdata/aggregate, line 301 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qVlydOdV0Gsr9z25P:-K5qmJTxmafOmd--oZ31:-645842991) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/testdata/aggregate#L301)):*
Also, add some more complex error cases.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[sql/group.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkOYnMoMtxcjsprY5:-K5qmmZxmzB02E5Lzs9O:1159103) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L164)):*
Doing so will be equivalent as the array will be heap allocated. One advantage to the current approach is that we could do: `scratch = encoded[0:0]` so that we take advantage of any growth in the slice on subsequent iterations. Doing that, we could actually initialize `scratch` as `var scratch []byte`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[sql/group.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkOYnMoMtxcjsprY5:-K5qo_rG8tCs8xvlQztO:-2103718613) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L164)):*
Sounds good; let's do that then (`var scratch []byte`). For reference, from @petermattis via slack:

> The only time an array stays on the stack is if it is used solely within a function or within functions that have been inlined.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5rRt-4MAO8QcKj6BI7:918089336) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
I renamed the bound `err` to `invalidAgg` to make it clearer what it is and why it isn't checked first. Also changed body of `checkAggregateExprs` to only return a non-nil error when `aggregated` is also true. 

---

*[sql/group.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qjAJXlY36Ij4mZH9A:-K5rRLSKNXlug1xHrhd0:-836698680) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L55)):*
There's a longer-form comment on it already, so I just shorted this one to reduce duplication.

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K5rOFUhxCbb9jsaPplM:-1892928839) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
I was doing that before but undid it when I started deferring initialization of `values` until `Next` calls `computeAggregates`, which made this NPE since `Columns` is called earlier.

---

*[sql/group.go, line 278 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkyc0Ps8I-C4TUNvc:-K5rSfQgkQ1wVZuq2QlW:-162496957) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L278)):*
renamed it `create` (`new` is a reserved word, right?)

---

*[sql/group.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qlgfSd51uzbtPr2Zt:-K5rSnlSn_EVwMnkqq72:-578977779) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L399)):*
put one below: `// TODO(davidt): consider other ways of comparing expression trees.`

---

*[sql/group.go, line 471 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qlvMTsxgu223XIVWR:-K5rEPA6pmRNRUJJlyI1:-196031907) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L471)):*
It happens in the case where a bucket is rendered without ever having a row added, eg `SELECT MAX(a) FROM t WHERE false;` is supposed to render 1 row (of NULL).

---

*[sql/group.go, line 528 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qm6hPk5aZNsGrz5wa:-K5rTJPRrzK6IBypJzJV:-970823428) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L528)):*
ah, i don't think it ever is uninitialized actually once I got everything else working correctly -- i'll take this out for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5rZNdhALkE45cz0PRV:-1144632006) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
Can you name it invalidAggErr?

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K5rZROksFe4eaI9S_lP:843033699) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
What if you also called `computeAggregates` on a call to Columns?

---

*[sql/group.go, line 278 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkyc0Ps8I-C4TUNvc:-K5rZ_QlREuWyFmCSfzL:-64351896) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L278)):*
not sure, but create is fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[sql/group.go, line 141 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfVrUtHjjTCn09TCd:-K5rfVrUtHjjTCn09TCe:660613550) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L141)):*
doesn't really matter, but you have used `TODO(dt)` elsewhere in this PR.

---

*[sql/group.go, line 325 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfkZ9V8axcbZhj_XQ:-K5rfkZ9V8axcbZhj_XR:-1573131589) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L325)):*
why can't this happen?

---

*[sql/group.go, line 368 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rgLy5OKfg9BXFgxYz:-K5rgLy5OKfg9BXFgxZ-:1965958265) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L368)):*
what speaks against naming this `err`? I expect a `bool` with the current name.

---

*[sql/group.go, line 403 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rg5fPHcfgQaoHGN7W:-K5rg5fPHcfgQaoHGN7X:-209346506) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L403)):*
Can you add to the comment a simple example of something that would be valid and something that wouldn't?

---

*[sql/group.go, line 518 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfKSIflXHb1b4VlaB:-K5rfKSIflXHb1b4VlaC:775231564) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L518)):*
is a line `var _ aggregateImpl = &identAggregate` missing? Also, I'm not sure what the "ident aggregate" is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[sql/group.go, line 325 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfkZ9V8axcbZhj_XQ:-K5rjCah3R239wKfoPul:-1881475605) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L325)):*
No aggregate functions take more than one argument. Type checking has already run on these expressions, which would have returned type errors for aggregate function calls with incorrect arities; if an aggregate function of the wrong arity gets here, something has gone really wrong.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/group.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qTHiOf8XdmAM-7blv:-K5s8lCea-lljOuvReTy:1700997001) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/group.go#L57)):*
Ditto.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K62A93-CLc0iu36A5jP:1651284843) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
hmm, might callers be surprised if `Columns()` started being expensive / having side-effects? I could also just go back to initializing `values` earlier (so this could delegate to it) and bring back the explicit bool indicating whether the rows had been initialized. 

---

*[sql/group.go, line 141 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfVrUtHjjTCn09TCd:-K62A5zjqIK-EazfnkIz:-716259727) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L141)):*
whoops, old habits. fixed.

---

*[sql/group.go, line 325 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfkZ9V8axcbZhj_XQ:-K62B3sNCmaaEjaH8EZY:-510179446) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L325)):*
Just copy-pasted that into a comment to the same effect.

---

*[sql/group.go, line 368 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rgLy5OKfg9BXFgxYz:-K626IO535YGGHI6DEMN:-540666299) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L368)):*
Since it is only meaningful, and thus only checked, when aggregation is being used, I wanted to avoid a name that suggested it could be used to return arbitrary errors, since otherwise those might mistakenly be ignored when aggregation is not in use.

---

*[sql/group.go, line 518 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rfKSIflXHb1b4VlaB:-K629x__QjdyvCkWJQST:-1988928154) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L518)):*
Added the interface check and a comment explaining what it is and why we need it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/group.go, line 368 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rgLy5OKfg9BXFgxYz:-K62DWiME-ndhlgNZcwB:334469476) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L368)):*
that makes sense. `aggrErr`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/group.go, line 368 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5rgLy5OKfg9BXFgxYz:-K63mqohg-aeZ3pe_knP:-469593337) ([raw file](https://github.com/cockroachdb/cockroach/blob/0975a50ad1951014885480dc63feca12755a2f26/sql/group.go#L368)):*
done.

---

*[sql/testdata/aggregate, line 301 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qVlydOdV0Gsr9z25P:-K63m4wlQTcXjWQCG7eu:-1091072372) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4fa2904c7a457e1a421320e537420cf363b62d2/sql/testdata/aggregate#L301)):*
I believe it does. DISTINCT is per-bucket, as far as I can tell.
Prefixed the `seen` keys with the bucket key to implement this, and added a test case that ensures same output as postgres.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/testdata/aggregate, line 10 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K64P0wTw9OVUiDgN7Gm:-K64P0wTw9OVUiDgN7Gn:1643273166) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c2509546b0331d777136746e40a321974fcf2d/sql/testdata/aggregate#L10)):*
I believe you can break this into multiple lines. The logic test parses the statement as everything up to the next blank line. That's why multi-line `CREATE TABLE` works.

---

*[sql/testdata/aggregate, line 44 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K64PtQfgwpx5-TV1lPj:-K64PtQfgwpx5-TV1lPk:1569757570) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c2509546b0331d777136746e40a321974fcf2d/sql/testdata/aggregate#L44)):*
Do you have a test which uses a table qualified name in one part of the query and an unqualified name in another? Something like:

```
SELECT UPPER(kv.s) FROM kv GROUP BY s
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-:-K64PylAw9GhfOLmRq_0:676834233)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/testdata/aggregate, line 44 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K64PtQfgwpx5-TV1lPj:-K64QdrA7MLrqRqEKLni:1068625008) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c2509546b0331d777136746e40a321974fcf2d/sql/testdata/aggregate#L44)):*
added (with one for each variant of one side qualified, other not).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K6GLGwW3COZFKUmUhJM:-24586492) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
Yeah, maybe initializing `values` earlier is the way to go.

---

*[sql/testdata/aggregate, line 106 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K6GJS7K2qs3JrbPShDk:-K6GJS7LplOeIknmvWob:1880355411) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c2509546b0331d777136746e40a321974fcf2d/sql/testdata/aggregate#L106)):*
nit: it's nice to line these up

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/group.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qkIxx_pVhHG7UvIJb:-K6UxGY7ZdLNluug05qj:-469593337) ([raw file](https://github.com/cockroachdb/cockroach/blob/a642f4816d13aeab1f9d94d5149acc2e6fdd3c29/sql/group.go#L118)):*
done.

---

*[sql/testdata/aggregate, line 106 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K6GJS7K2qs3JrbPShDk:-K6UxnLJ68qGKyCVmEkC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c2509546b0331d777136746e40a321974fcf2d/sql/testdata/aggregate#L106)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented on sql/group.go:
```diff
@@ -94,68 +111,108 @@ func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
-				return false
+		aggregatedValues, groupedValues := values[:len(n.funcs)], values[len(n.funcs):]
+
+		//TODO(dt): optimization: skip buckets when underlying plan is ordered by grouped values.
```
nit: space after `//` (here and elsewhere)

tamird commented on sql/group.go:
```diff
@@ -356,27 +455,65 @@ func (a *aggregateFunc) Add(d parser.Datum) error {
+	}
+
+	datum, err := found.result()
+
```
some weird spaces here

tamird commented on sql/group.go:
```diff
@@ -24,69 +24,86 @@ import (
-	if len(funcs) == 0 {
+
+	// Determine if aggregation is being performed and, if so, if it is valid,
+	// i.e. ever render expression either aggregates all qvalues or appears in
```
s/ever/every/

tamird commented on sql/group.go:
```diff
@@ -24,69 +24,86 @@ import (
+	// value added to them for a bucket) to provide grouped-by values for each bucket.
+	// After extraction, group.render will be entirely rendered from aggregateFuncs,
+	// and group.funcs will contain all the functions which need to be fed values.
+
```
nit: remove this empty line

tamird commented on sql/group.go:
```diff
@@ -24,69 +24,86 @@ import (
+	}
+
+	// Loop over the render expressions and extract any aggregate functions --
+	// qvalues are also replaced (with identAggregates, which just returns the last
```
s/returns/return/

tamird commented on sql/group.go:
```diff
@@ -94,68 +111,108 @@ func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
+				return
 			}
 		}
+		scratch = encoded[0:0]
```
shouldn't this be `scratch = scratch[:0]`?

petermattis replied:
`encoded` might have been reallocated (grown) at L171 by `encodeDTuple`, so better to use the larger temporary buffer at this point. Though this could be `encoded[:0]`.

tamird replied:
Good point.

tamird commented on sql/group.go:
```diff
@@ -94,68 +111,108 @@ func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
-		if n.err != nil {
-			return false
+	n.values.rows = make([]parser.DTuple, 0, len(n.buckets))
+
```
nit: remove this line? this initialization is attached to the next for-iteration

tamird commented on sql/group.go:
```diff
@@ -94,68 +111,108 @@ func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
+		n.currentBucket = k
+
+		row := make(parser.DTuple, 0, len(n.render))
+
```
nit: remove this line? this initialization is attached to the next for-iteration

tamird commented on sql/group.go:
```diff
@@ -259,93 +316,135 @@ func (v *extractAggregatesVisitor) Visit(expr parser.Expr, pre bool) (parser.Vis
 }
 
-func (a *aggregateFunc) Add(d parser.Datum) error {
+func (a *aggregateFunc) add(bucket string, d parser.Datum) error {
```
this signature forces a number of string-to-`[]byte`-to-string conversions that can be avoided

dt replied:
so, it is only converted back to a `[]byte` for the `seen` distinct case -- and it _is_ used as a string (since map keys can't be []byte, right?). deferring the `sting()` conversion would mean doing twice -- once here and once in the caller... I guess we could just pass both? ugh.

dt replied:
Actually, I think in the case of `seen` distinct check case, I _do_ want to a *copy* of the bucket key bytes to hand to `encodeDatum`, since it will be appending the value to it, but I don't want to change the bucket key as used below.

tamird replied:
i think the copy isn't strictly necessary (because encoding mutates the underlying array, but not the provided slice). this is fine as is, since this is subtle enough to not be worth the optimization.

tamird commented on sql/group.go:
```diff
@@ -356,27 +455,65 @@ func (a *aggregateFunc) Add(d parser.Datum) error {
+
+func (a *aggregateFunc) Eval(ctx parser.EvalContext) (parser.Datum, error) {
+	found, ok := a.buckets[a.group.currentBucket]
+
```
this one too

tamird commented on sql/group.go:
```diff
@@ -384,29 +521,53 @@ var _ aggregateImpl = &countAggregate{}
 var _ aggregateImpl = &sumAggregate{}
+var _ aggregateImpl = &identAggregate{}
+
+// In order to render the unaggregated (ie grouped) fields, during aggregation,
```
s/ie/i.e./

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/group.go, line 54 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-54:-K6eSffzJbicR4RG1Wr7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L54)):*
Done.

---

*[sql/group.go, line 69 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-69:-K6eSgwAxmc4pcSOmcoC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L69)):*
Done.

---

*[sql/group.go, line 73 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-73:-K6eShy66LikHmTg45Bz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L73)):*
Done.

---

*[sql/group.go, line 168 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-168:-K6eSkSsnjwe9T12K_EY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L168)):*
Done.

---

*[sql/group.go, line 185 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-185:-K6eSmCpovVoiK0TrLq6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L185)):*
Done.

---

*[sql/group.go, line 199 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-199:-K6eSnC9aoQa3QkAdzPb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L199)):*
Done.

---

*[sql/group.go, line 204 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-204:-K6eSnThqiWpC-DHeM5E:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L204)):*
Done.

---

*[sql/group.go, line 481 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-481:-K6eSqtbTS-SKNcCTOMY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L481)):*
Done.

---

*[sql/group.go, line 487 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-487:-K6eSrmEans5jgDRRfhL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L487)):*
Done.

---

*[sql/group.go, line 526 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-526:-K6eSsLSjocz9FA_S0hR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L526)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/group.go, line 54 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-54:-K6evZiSq6W2AYdqFDMX:542681710) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L54)):*
"either aggregates qvalues or ..." is this actually true? it should be possible to aggregate arbitrary non-aggregate expressions

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/group.go, line 54 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-54:-K6iRLRYRPvHxWZ3VBfD:-1874911577) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L54)):*
yeah, if there are no qvalues, it is vacuously true that it aggregates "all qvalues".

---

*[sql/group.go, line 445 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494#-K5qLPj4hCOjqB3X-vLY-r8-445:-K6eWRdwdYba6zzLcRBV:-1173100375) ([raw file](https://github.com/cockroachdb/cockroach/blob/65653f55d89c2b16d7f7b75d5f357b4648d23e19/sql/group.go#L445)):*
I'm not sure that i've seen an approach I like better than the current -- supposedly calls to `string(myByteSlice)` are optimized by the compiler if they are only temporary (eg used only locally as map keys)  but relying on that feels icky. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3494)*
<!-- Sent from Reviewable.io -->
